Enable support for simd directives

kiranchandramohan opened the pull request:
This commit enables support for the following simd directives. These
directives can be used to control vectorisation performed by LLVM. It
works by passing vectorisation metadata to LLVM.
Note that these directives do not control any vectorisation performed
by flang itself.
1) !dir$ ivdep (ignore mem-dep and vectorise)
2) !$omp simd (requires -fopenmp, ignore mem-dep and vectorise)
3) !dir$ vector always (ignore cost and vectorise)
4) !dir$ novector

gklimowicz reviewed (approved):
This passed my tests on OpenPOWER for LLVM 9 and 10.

bryanpkc commented on test/f90_correct/novector.f90:
```diff
@@ -0,0 +1,25 @@
+! RUN: %flang -S -emit-llvm %s -o - | FileCheck %s -check-prefix=DISABLE-METADATA
+! RUN: %flang -Hx,59,2 -S -emit-llvm %s -o - | FileCheck %s -check-prefix=METADATA
```
The FileCheck prefix here is confusing. I suggest renaming it to `IGNORE-DIRECTIVES`.

bryanpkc replied:
The `VECTOR` one below is confusing too. The test is actually trying to confirm that vector instructions *are not* generated.

kiranchandramohan replied:
Done.

bryanpkc commented on test/f90_correct/novector.f90:
```diff
@@ -0,0 +1,25 @@
+end subroutine
+! DISABLE-METADATA: !"llvm.loop.vectorize.enable", i1 false
+! METADATA-NOT: !"llvm.loop.vectorize.enable", i1 false
+! DISABLE_DIRECTIVE: remark: vectorized loop
```
This FileCheck pattern is not used.

kiranchandramohan replied:
Done.

bryanpkc commented on test/f90_correct/vector.f90:
```diff
@@ -0,0 +1,21 @@
+! RUN: %flang -S -emit-llvm %s -o - | FileCheck %s -check-prefix=METADATA
+! RUN: %flang -Hx,59,2 -S -emit-llvm %s -o - | FileCheck %s -check-prefix=ENABLE-METADATA
```
This FileCheck prefixes for the negative tests are confusing.

kiranchandramohan replied:
Done.

bryanpkc commented on tools/shared/pragma.c:
```diff
@@ -647,10 +648,34 @@ do_sw(void)
       bset(DIR_OFFSET(currdir, x[19]), 0x18); /* notransform | norecog */
-    else
-      bclr(DIR_OFFSET(currdir, x[19]), 0x18);
+      bclr(DIR_OFFSET(currdir, x[19]), 0x400);
```
This code seems to be repurposing `XBIT(19, 0x400)`. Could you please update `tools/flang2/docs/xflag.n`?

kiranchandramohan replied:
Done.

bryanpkc reviewed (changes requested):
Could you move all tests from `test/f90_correct/` to a new directory, e.g. `test/directives/`?

Please add the standard license boilerplate in the test cases.

bryanpkc commented on tools/flang2/flang2exe/cgmain.cpp:
```diff
@@ -1557,15 +1556,20 @@ schedule(void)
+    BIH_NODEPCHK(bih) = !flg.depchk;
+    if(XBIT(19, 0x18))
+      BIH_NOSIMD(bih) = true;
+    else if(XBIT(19, 0x400))
```
Please insert a space between `if` and the left parenthesis, on both lines 1561 and 1563.

kiranchandramohan replied:
Done.

kiranchandramohan commented on test/f90_correct/omp_simd_clause.f90:
```diff
@@ -0,0 +1,21 @@
+  integer, pointer :: myarr1(:)
+  integer, pointer :: myarr2(:)
+  integer :: ub
+  
```
Done.

kiranchandramohan commented on test/f90_correct/omp_simd.f90:
```diff
@@ -0,0 +1,19 @@
+  integer, pointer :: myarr1(:)
+  integer, pointer :: myarr2(:)
+  integer :: ub
+  
```
Done.

kiranchandramohan commented on test/f90_correct/ivdep.f90:
```diff
@@ -0,0 +1,19 @@
+  integer, pointer :: myarr1(:)
+  integer, pointer :: myarr2(:)
+  integer :: ub
+  
```
Done.

kiranchandramohan commented:
> Could you move all tests from `test/f90_correct/` to a new directory, e.g. `test/directives/`?

Thanks @bryanpkc for the comments. I have moved tests to a new directory and addressed all other changes.

> Please add the standard license boilerplate in the test cases.

This one I did not do since the license is not required in test cases in llvm. For e.g. consider the test below.
https://github.com/llvm/llvm-project/blob/master/llvm/test/Analysis/BasicAA/128-bit-ptr.ll

bryanpkc commented on tools/flang2/flang2exe/bih.h:
```diff
@@ -63,6 +63,7 @@ typedef struct {
       unsigned parsect : 1;   /* bih belongs to a parallel section */
       unsigned ujres : 1;     /* bih contains ujresidual start & count info */
       unsigned simd : 1;       /* bih contains simd code */
+      unsigned nosimd : 1;       /* bih contains simd code */
```
Should the comment say `bih does not contain simd code`?

kiranchandramohan replied:
Done.

bryanpkc reviewed (changes requested):
`test/f90_correct/vector_directive{1,2,3}.f90` should be moved to `test/directives/`.

kiranchandramohan commented:
Thanks @bryanpkc, i have addressed the last comment too.

@shivaramaarao can you have a look too?

@sscalpone has a mail asking about status of OpenMP in classic flang. This patch adds support for !$omp simd which can affect the status.
http://lists.llvm.org/pipermail/flang-dev/2020-October/000577.html

bryanpkc reviewed (approved):
Thanks @kiranchandramohan. The branch LGTM.

bryanpkc commented:
Sorry, `test/f90_correct/vector_directive{1,2,3}.f90` still haven't been moved to `test/directives/`. @kiranchandramohan

kiranchandramohan commented:
> Sorry, `test/f90_correct/vector_directive{1,2,3}.f90` still haven't been moved to `test/directives/`. @kiranchandramohan

Ooops, moved them also. Thanks for pointing out.

shivaramaarao reviewed (approved):
LGTM
